nemo/collections/asr/parts/submodules/tdnn_attention.py

```diff
@@ -0,0 +1,259 @@
+            self.feat_in += feat_in
+        elif self.pool_mode == 'tap':
+            self.feat_in = feat_in
+        else:
+            raise ValueError("pool mode for stats must be either tap or xvector based")
+
+    def forward(self, encoder_output, length=None):
+        mean = encoder_output.mean(dim=-1)  # Time Axis
+        if self.pool_mode == 'xvector':
+            std = encoder_output.std(dim=-1)
+            pooled = torch.cat([mean, std], dim=-1)
+        else:
+            pooled = mean
+        return pooled
+
+
+def lens_to_mask(lens, max_len, device=None):
+    """
+    outputs masking labels for list of lengths of audio features, with max length of any 
+    mask as max_len
+    input:
+        lens: list of lens
+        max_len: max length of any audio feature
+    output:
+        mask: masked labels
+        num_values: sum of mask values for each feature (useful for computing statistics later)
+    """
+    lens_mat = torch.arange(max_len).to(device)
+    lens = lens * max_len
+    mask = lens_mat[:max_len].unsqueeze(0) < lens.unsqueeze(1)
+    mask = mask.unsqueeze(1)
+    num_values = torch.sum(mask, dim=2, keepdim=True)
+    return mask, num_values
+
+
+def get_statistics_with_mask(x, m, dim=2, eps=1e-10):
+    """
+    compute mean and standard deviation of input(x) provided with its masking labels (m)
+    input:
+        x: feature input 
+        m: averaged mask labels 
+    output:
+        mean: mean of input features
+        std: stadard deviation of input features
+    """
+    mean = torch.sum((m * x), dim=dim)
+    std = torch.sqrt((m * (x - mean.unsqueeze(dim)).pow(2)).sum(dim).clamp(eps))
+    return mean, std
+
+
+class TDNN_Module(nn.Module):
+    """
+    Time Delayed Neural Module (TDNN) - 1D
+    input:
+        inp_filters: input filter channels for conv layer
+        out_filters: output filter channels for conv layer
+        kernel_size: kernel weight size for conv layer
+        dilation: dilation for conv layer
+        stride: stride for conv layer
+        padding: padding for conv layer (default None: chooses padding value such that input and output feature shape matches)
+    output:
+        tdnn layer output 
+    """
+
+    def __init__(self, inp_filters, out_filters, kernel_size=1, dilation=1, stride=1, padding=None):
+        super(TDNN_Module, self).__init__()
+        if padding is None:
+            padding = get_same_padding(kernel_size, stride=stride, dilation=dilation)
+
+        self.conv_layer = nn.Conv1d(
+            in_channels=inp_filters,
+            out_channels=out_filters,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            padding=padding,
+        )
+
+        self.activation = nn.ReLU()
+        self.bn = nn.BatchNorm1d(out_filters)
+
+    def forward(self, x, length=None):
+        x = self.conv_layer(x)
+        x = self.activation(x)
+        return self.bn(x)
+
+
+class SE_Module(nn.Module):
```

Comment:
Why reimplement SEBlock from jasper.py?? Revert this and use that. If there is really no way to use that, only then rename this to `TDNNSEBlock` and use it in `TDNNSEModule`

Reply:
`jasper.py` uses linear layer and also order is different. Observed conv1d layers with kernel size=1 gave better numbers and also converged faster

Reply:
What's different about order ? And conv 1d with pointwise kernel is mathematically equivalent to dense layer

Reply:
They do have different initializations though. Hmm fine just rename it for now

Reply:
it doesn't; perform batch norm and also I take care of length masking